WordPress/Classes/ViewRelated/Activity/ActivityListViewModel.swift

```diff
@@ -40,17 +40,41 @@ enum ActivityListViewModel {
         case .loading, .error:
             return .Empty
         case .ready(let activities):
-            let rows = activities.map({ activity in
+            var rows = activities.filter({ activity in
+                !activity.isDiscarded
+            }).map({ activity in
                 return ActivityListRow(
                     activity: activity,
                     action: { (row) in
                         presenter.presentRewindFor(activity: activity)
                     }
                 )
             })
-            return ImmuTable(sections: [
-                ImmuTableSection(rows: rows)
-            ])
+
+            guard rows.count > 0 else {
+                return ImmuTable(sections: [])
+            }
+
+            var tableSections = [ImmuTableSection]()
+
+            let firstRow = rows.removeFirst()
+            var currentSection: (date: String, rows: [ImmuTableRow]) = (firstRow.activity.published.longUTCStringWithoutTime(), [firstRow])
+
+            for row in rows {
+                if row.activity.published.longUTCStringWithoutTime() == currentSection.date {
+                    currentSection.rows.append(row)
+                } else {
+                    tableSections.append(ImmuTableSection(headerText: currentSection.date,
+                                                          optionalRows: currentSection.rows,
+                                                          footerText: nil)!)
+                    currentSection = (row.activity.published.longUTCStringWithoutTime(), [row])
+                }
+            }
+            tableSections.append(ImmuTableSection(headerText: currentSection.date,
+                                                  optionalRows: currentSection.rows,
+                                                  footerText: nil)!)
+
+            return ImmuTable(sections: tableSections)
```

Comment:
This was a bit hard to follow. I'd try to move the algorithm for grouping somewhere else, maybe an `Array` extension that returns a `[String: [Element]]`.

I have to go for today and I don't have a prettier algorithm, but here's a stab at a generic version that can be unit tested:

```swift
extension Array {
    func sortedGroup<Value: Equatable>(keyPath: KeyPath<Element, Value>) -> [(Value, [Element])] {
        var currentValue: Value?
        var currentGroup = [Element]()
        var result = [(Value, [Element])]()
        forEach { (element) in
            let value = element[keyPath: keyPath]
            if currentValue != value {
                if let currentValue = currentValue {
                    result.append((currentValue, currentGroup))
                }
                currentValue = value
                currentGroup = []
            }
            currentGroup.append(element)
        }
        if let currentValue = currentValue {
            result.append((currentValue, currentGroup))
        }
        return result
    }
}
```

In any case, you can probably use `ImmuTable(optionalSections:)` and avoid those `!`

Reply:
Actually, I don't thin there's a need for the `optionalRows` variant of `ImmuTableSection`. The `currentSection.rows` is an array of non-optional rows, so you can use the regular section.

Reply:
😍 will do!